numpyro/mcmc.py

```diff
@@ -428,3 +441,303 @@ def single_chain_mcmc(rng, init_params):
         return samples_flat
     else:
         raise ValueError('sampler: {} not recognized'.format(sampler))
+
+
+# ========= Higher Level API ========= #
+
+
+class MCMCKernel(ABC):
+    """
+    Defines the interface for the Markov transition kernel that is
+    used for :class:`~numpyro.mcmc.MCMC` inference.
+
+    :param random.PRNGKey rng: Random number generator key to initialize
+        the kernel.
+    :param int num_warmup: Number of warmup steps. This can be useful
+        when doing adaptation during warmup.
+    :param tuple init_params: Initial parameters to begin sampling. The type must be consistent
+            with the input type to `potential_fn`.
+    :param model_args: Arguments provided to the model.
+    :param model_kwargs: Keyword arguments provided to the model.
+    """
+    @abstractmethod
+    def init(self, rng, num_warmup, init_params, model_args, model_kwargs):
+        raise NotImplementedError
+
+    @abstractmethod
+    def sample(self, state):
+        """
+        Given the current `state`, return the next `state` using the given
+        transition kernel.
+
+        :param state: Arbitrary data structure representing the state for the
+            kernel. For HMC, this is given by :data:`~numpyro.mcmc.HMCState`.
+        :return: Next `state`.
+        """
+        raise NotImplementedError
+
+
+class HMC(MCMCKernel):
+    """
+    Hamiltonian Monte Carlo inference, using fixed number of steps with
```

Comment:
nit: maybe `number of steps` -> `trajectory length` (which is more correct, especially with adaptation scheme)?